scipy/fft/tests/test_basic.py

```diff
@@ -19,251 +32,382 @@ def fft1(x):
+            op_tr = op(xp_test.permute_dims(x, axes=a),
+                       s=shape[:2], axes=(0, 1))
+            tr_op = xp_test.permute_dims(op(x, s=shape[:2], axes=a[:2]),
+                                         axes=a)
+            assert_allclose(op_tr, tr_op)
+
+    @skip_if_array_api_gpu
+    @array_api_compatible
+    def test_all_1d_norm_preserving(self, xp):
         # verify that round-trip transforms are norm-preserving
-        x = random(30)
-        x_norm = np.linalg.norm(x)
-        n = x.size * 2
+        x = xp.asarray(random(30))
+        xp_test = array_namespace(x)
+        x_norm = xp_test.linalg.vector_norm(x)
+        n = size(x) * 2
         func_pairs = [(fft.fft, fft.ifft),
                       (fft.rfft, fft.irfft),
                       # hfft: order so the first function takes x.size samples
                       #       (necessary for comparison to x_norm above)
                       (fft.ihfft, fft.hfft),
                       ]
         for forw, back in func_pairs:
-            for n in [x.size, 2*x.size]:
+            for n in [size(x), 2*size(x)]:
                 for norm in ['backward', 'ortho', 'forward']:
                     tmp = forw(x, n=n, norm=norm)
                     tmp = back(tmp, n=n, norm=norm)
-                    assert_array_almost_equal(x_norm,
-                                              np.linalg.norm(tmp))
-
-    @pytest.mark.parametrize("dtype", [np.float16, np.float32, np.float64,
-                                       np.longdouble])
-    def test_dtypes(self, dtype):
-        # make sure that all input precisions are accepted
-        x = random(30).astype(dtype)
+                    assert_allclose(x_norm, xp_test.linalg.vector_norm(tmp))
+
+    @skip_if_array_api
+    def test_float16(self):
+        x = random(30).astype(np.float16)
         assert_array_almost_equal(fft.ifft(fft.fft(x)), x)
         assert_array_almost_equal(fft.irfft(fft.rfft(x)), x)
         assert_array_almost_equal(fft.hfft(fft.ihfft(x), len(x)), x)
 
-
+    @array_api_compatible
+    def test_float32(self, xp):
+        x = xp.asarray(random(30), dtype=xp.float32)
+        _assert_allclose = set_assert_allclose(xp)
+        _assert_allclose(fft.ifft(fft.fft(x)),
+                         xp.asarray(x, dtype=xp.complex64),
+                         rtol=1e-4, atol=0)
+        _assert_allclose(fft.irfft(fft.rfft(x)), x,
+                         rtol=1e-4, atol=0)
+        _assert_allclose(fft.hfft(fft.ihfft(x), x.shape[0]), x,
+                         rtol=1e-4, atol=0)
+
+    @array_api_compatible
+    def test_float64(self, xp):
+        x = xp.asarray(random(30), dtype=xp.float64)
+        _assert_allclose = set_assert_allclose(xp)
+        _assert_allclose(fft.ifft(fft.fft(x)),
+                         xp.asarray(x, dtype=xp.complex128))
+        _assert_allclose(fft.irfft(fft.rfft(x)), x)
+        _assert_allclose(fft.hfft(fft.ihfft(x), x.shape[0]), x)
+
+
```

Comment:
do we want to test `longdouble` here?

Reply:
Will leave that for @rgommers , but I don't think `longdouble` is implemented for anything that's not NumPy!

Reply:
Yeah, I'm more asking whether we should test it for just NumPy, like `float16`.

Reply:
It's already tested a few lines down (L402), right? No need to overdo it on `longdouble` I'd say.

Reply:
It's included in `test_fft_with_order`. I'm talking about how I have replaced the `test_dtypes` test, but am no longer testing `longdouble` in the replacements (like `test_float16`).

Reply:
Got it. I had a close look, and I'd rework it a bit and keep the tests parametrized: https://github.com/scipy/scipy/commit/e3b6cc84c84a02d823fec6a62d0d6e68808dd1a7. This puts back `longdouble, and also accounts for PyTorch checking dtype matches (`numpy.testing` functions don't do that), and increasing the test precision for float64. 

Also, there was a small bug in `_basic.py`: https://github.com/scipy/scipy/commit/24164ed2cb3c0ea364502314699b202e20aa9faa

Reply:
Thanks, this looks a lot better! There may be other places where similar improvements can be made, but I think this was the only test that I "replaced" like this. And thanks for spotting the bug; does the refactor to use those `execute` functions look good otherwise? If so, a similar approach can probably be applied to the real transforms.

Reply:
That refactor looks quite good to me, nice work. I'd probably go for factoring out the argument validation too: https://github.com/scipy/scipy/commit/eb796da7c29384ccaf979c8b1e53097b373f2caf. That way the clearly common part is separated out, and it's easier to read the logic of 1-D vs. n-D.

Reply:
I've cherry-picked those three commits.

Reply:
I've just applied a similar refactor to `_realtransforms.py`: https://github.com/scipy/scipy/pull/19005/commits/a889b7da8b66aff46fa66f45199eb96e8c26b7c6.